kemeny_sourcemash.tex

```diff
@@ -59,15 +59,41 @@ \subsection{Development: localhost}
 
-\subsection{Subsection 1}
+\subsection{Categorization Algorithm}
 
-When
+Because we need articles to group together by category, it was import that our categorization algorithm generate categories with a decent number of overlapping categories. In addition, it was important that the algorithm work quickly because it would need to process many articles 24/7 to stay up-to-date with the latest news.
 
-\subsection{Subsection 2}
+To help with the first concern, we needed a way to normalize our category data. This meant that capitalization, punctuation, and pluralization should all be consistent for each instance of a given category. Upon researching state-of-the-art categorization techniques, we came across the CommunityCluster algorithm\cite{Grineva}, which uses Wikipedia lookups to generate possible keywords from an article. By using Wikipedia articles, categories would be normalized by converting n-grams from a source article into official Wikipedia article titles.
 
-When
+For the second concern, we sped up the CommunityCluster algorithm by only running it on the n-most-common ngrams in each feed article. This would require less lookups using the Wikipedia API.
+
+The CommunityCluster algorithm computes relatedness between two wikipedia articles by counting the number of overlapping links between two articles per total number of links. This relatedness score becomes the edge weight between the articles. Once all articles and their edges are added the graph, the graph generates clusters and picks out the densest clusters as the selected keywords.
+
+In the end, our algorithm works as follows:
+
+\begin{enumerate}
+
+\item Generate bag-o-words for article (a), weighting counts for n-grams from the title or longer than a unigram more heavily.
+\item From the word counts generated in the bag-o-words, select the 20 most common ngrams as the keyword candidates.
+\item For each keyword candidate (c), search Wikipedia for whether (c) exists as a wikipedia article (w).
+  \begin{enumerate}
+  \item If the article exists but is ambiguous (i.e. has a Wiki disambiguation page), map the article to all related (w) linked to from the disambiguation page.
+  \item Otherwise, map (c) to itself as the only (w).
+  \end{enumerate}
+\item Memoize the cross-wiki links in each (w) to quickly compute relatedness scores later.
+\item For each (c) that only mapped to one (w), store (w) in the list of assigned articles.
+\item For each ambiguous (c), sum the relatedness scores between each (w) and the list of assigned articles in Step 5. Add (w) with the greatest summed relatedness score to the list of assigned articles.
+\item Generate the graph (g), where each (w) in the assigned articles list is a vertex and the relatedness between itself and another article is the edge weight.
+\item Perform the Louvain method to isolate the communities in (G).
+\item Extract the densest communities' keywords to be used as the keywords for (a).
+\end{enumerate}
+
+
```

Comment:
Future Extensions? Something like:

Possible mechanisms for enhancing this algorithm include post-processing category data to normalize similar categories, disallowing nested categories (e.g. choose Google Maps over Google on an item instead of including both), etc etc...........